src/Dictionaries/FlatDictionary.cpp

```diff
@@ -151,6 +152,108 @@ ColumnPtr FlatDictionary::getColumn(
+
+    size_t attribute_index = dict_struct.attribute_name_to_index.find(attribute_name)->second;
+    const auto & attribute = attributes[attribute_index];
+
+    bool is_attribute_nullable = attribute.is_nullable_set.has_value();
+    ColumnUInt8::MutablePtr col_null_map_to;
+    ColumnUInt8::Container * vec_null_map_to = nullptr;
+
+    if (is_attribute_nullable)
+    {
+        col_null_map_to = ColumnUInt8::create(size, false);
+        vec_null_map_to = &col_null_map_to->getData();
+    }
+
+    auto type_call = [&](const auto & dictionary_attribute_type)
+    {
+        using Type = std::decay_t<decltype(dictionary_attribute_type)>;
+        using AttributeType = typename Type::AttributeType;
+        using ValueType = DictionaryValueType<AttributeType>;
+        using ColumnProvider = DictionaryAttributeColumnProvider<AttributeType>;
+
+        auto column = ColumnProvider::getColumn(dictionary_attribute, size);
+
+        if constexpr (std::is_same_v<ValueType, Array>)
+        {
+            auto * out = column.get();
+
+            getItemsShortCircuitImpl<ValueType, false>(
+                attribute,
+                ids,
+                [&](size_t, const Array & value, bool) { out->insert(value); },
+                default_mask);
+        }
+        else if constexpr (std::is_same_v<ValueType, StringRef>)
+        {
+            auto * out = column.get();
+
+            if (is_attribute_nullable)
+                getItemsShortCircuitImpl<ValueType, true>(
+                    attribute,
+                    ids,
+                    [&](size_t row, StringRef value, bool is_null)
+                    {
+                        (*vec_null_map_to)[row] = is_null;
+                        out->insertData(value.data, value.size);
+                    },
+                    default_mask);
+            else
+                getItemsShortCircuitImpl<ValueType, false>(
+                    attribute,
+                    ids,
+                    [&](size_t, StringRef value, bool) { out->insertData(value.data, value.size); },
+                    default_mask);
+        }
+        else
+        {
+            auto & out = column->getData();
+
+            if (is_attribute_nullable)
+                getItemsShortCircuitImpl<ValueType, true>(
+                    attribute,
+                    ids,
+                    [&](size_t row, const auto value, bool is_null)
+                    {
+                        (*vec_null_map_to)[row] = is_null;
+                        out[row] = value;
+                    },
+                    default_mask);
+            else
+                getItemsShortCircuitImpl<ValueType, false>(
+                    attribute,
+                    ids,
+                    [&](size_t row, const auto value, bool) { out[row] = value; },
+                    default_mask);
+        }
+
+        result = std::move(column);
+    };
+
+    callOnDictionaryAttributeType(attribute.type, type_call);
+
```

Comment:
@KochetovNicolai I met the problem: if the value column is integer, I have to add `result = result->filter(default_mask, found_count);` in this line. But if it is string type, with this statement, would result in error. My understanding is that the underneath metadata impl of filter for different types is not uniformed.

String example can be found in the test, and here is the integer example
```sql
CREATE TABLE source_table(id UInt64, v UInt64) ENGINE=TinyLog;
INSERT INTO source_table VALUES (0, 0), (1, 1);
CREATE DICTIONARY dict(id UInt64, v UInt64) PRIMARY KEY id SOURCE(CLICKHOUSE(TABLE 'source_table')) LIFETIME(MIN 0 MAX 0) LAYOUT(FLAT());
SELECT dictGetOrDefault('dict', 'v', id+1, intDiv(1, id)) FROM source_table;
```